src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
+        } else {
+            // leaf is on right
+            left_bytes = ark_ff::to_bytes!(&self.leaf_sibling_hash)?;
+            right_bytes = ark_ff::to_bytes!(&claimed_leaf_hash)?;
+        };
 
-            let mut prev = claimed_leaf_hash;
-            // Check levels between leaf level and root.
-            for &(ref hash, ref sibling_hash) in &self.path {
-                // Check if the previous hash matches the correct current hash.
-                if &prev != hash && &prev != sibling_hash {
-                    return Ok(false);
-                };
-                prev = hash_inner_node::<P::H>(parameters, hash, sibling_hash, &mut buffer)?;
+        let mut curr_path_node =
+            P::TwoToOneHash::evaluate_both(&two_to_one_hash_parameters, &left_bytes, &right_bytes)?;
+
+        let mut left_bytes;
+        let mut right_bytes;
+        // we will use `index` variable to track the position of path
+        let mut index = self.leaf_index;
+        index >>= 1;
+
+        // Check levels between leaf level and root
+        for level in (0..self.auth_path.len()).rev() {
+            // check if path node at this level is left or right
+            if index & 1 == 0 {
+                // curr_path_node is on the left
+                left_bytes = ark_ff::to_bytes!(&curr_path_node)?;
+                right_bytes = ark_ff::to_bytes!(&self.auth_path[level])?;
+            } else {
+                // curr_path_node is on the right
+                left_bytes = ark_ff::to_bytes!(&self.auth_path[level])?;
+                right_bytes = ark_ff::to_bytes!(&curr_path_node)?;
             }
+            // update curr_path_node
+            curr_path_node = P::TwoToOneHash::evaluate_both(
+                &two_to_one_hash_parameters,
+                &left_bytes,
+                &right_bytes,
+            )?;
+            index >>= 1;
+        }
 
-            if root_hash != &prev {
-                return Ok(false);
-            }
-            Ok(true)
-        } else {
-            Ok(false)
+        // check if final hash is root
+        if &curr_path_node != root_hash {
+            return Ok(false);
         }
+
+        Ok(true)
     }
 }
 
+/// Defines a merkle tree data structure.
+/// This merkle tree has runtime fixed height, and assumes number of leaves is 2^height.
+///
+/// TODO: add RFC-6962 compatible merkle tree in the future.
+/// For this release, padding will not be supported because of security concerns: if the leaf hash and two to one hash uses same underlying
+/// CRH, a malicious prover can prove a leaf while the actual node is an inner node. In the future, we can prefix leaf hashes in different layers to
+/// solve the problem.
 pub struct MerkleTree<P: Config> {
-    tree: Vec<Digest<P>>,
-    padding_tree: Vec<(Digest<P>, Digest<P>)>,
-    parameters: <P::H as FixedLengthCRH>::Parameters,
-    root: Option<Digest<P>>,
+    /// stores the non-leaf nodes in level order. The first element is the root node.
```

Comment:
```suggestion
    /// stores the non-leaf nodes in level order. The first element is the root node. 
    /// The ith nodes children are at indices 2*i, 2*i + 1
```

Reply:
just checked `left_child` and `right_child`. The ith nodes children are at indices `2*i + 1`, `2*i + 2`.